File: crm_dashboard/src/lib/lead-status.test.ts
import { describe, expect, it } from "vitest";
import { LEAD_STATUSES, type LeadStatus } from "./labels";
import { hasBeenConverted, isValidStatusTransition, statusTransitionOptions } from "./lead-status";

// This is the literal transition matrix from
// crm_be/internal/lead/usecase.go's validateStatusTransition, worked out
// by hand from the Go source (docs/phases/02-crm-core/td.md §5 + the
// documented "leaving lost" deviation) — not derived from
// isValidStatusTransition itself, so a bug that breaks BOTH the same way
// can't hide.
//
// Nothing leads to "new" (issue #139, ADR-015): "new" means "belum
// disentuh", a statement about history rather than a stage of work, so no
// row below lists it — not even lost's, which is why reopening goes to
// "contacted".
const EXPECTED_VALID: Record<LeadStatus, LeadStatus[]> = {
  new: ["contacted", "lost", "unqualified", "spam"],
  contacted: ["qualified", "lost", "unqualified", "spam"],
  qualified: ["contacted", "proposal", "lost", "unqualified", "spam"],
  proposal: ["qualified", "won", "lost", "unqualified", "spam"],
  won: ["proposal", "lost", "unqualified", "spam"],
  // "leaving lost" is a documented backend simplification: ANY main-path
  // status is valid, not just the one the lead was in before — there's
  // no cheap way to know "before" without activity history (crm_be #20
  // notes). unqualified/spam are also reachable directly from lost.
  lost: ["contacted", "qualified", "proposal", "won", "unqualified", "spam"],
  unqualified: [],
  spam: [],
};

describe("isValidStatusTransition", () => {
  it("matches the full matrix for every (from, to) pair — including same-status", () => {
    for (const from of LEAD_STATUSES) {
      for (const to of LEAD_STATUSES) {
        const expected = to === from ? to === "lost" : EXPECTED_VALID[from].includes(to);
        expect(
          isValidStatusTransition(from, to),
          `${from} -> ${to} expected ${expected}`
        ).toBe(expected);
      }
    }
  });

  it("unqualified and spam are final — no outgoing transition at all, not even to themselves", () => {
    for (const to of LEAD_STATUSES) {
      expect(isValidStatusTransition("unqualified", to)).toBe(false);
      expect(isValidStatusTransition("spam", to)).toBe(false);
    }
  });

  it("main path movement is exactly one step in either direction", () => {
    expect(isValidStatusTransition("qualified", "contacted")).toBe(true); // back
    expect(isValidStatusTransition("qualified", "proposal")).toBe(true); // forward
    expect(isValidStatusTransition("proposal", "qualified")).toBe(true); // back
    expect(isValidStatusTransition("won", "proposal")).toBe(true); // back
    expect(isValidStatusTransition("proposal", "new")).toBe(false); // two steps back
    expect(isValidStatusTransition("new", "won")).toBe(false); // skips ahead
  });
});

describe("nothing leads back to new (issue #139)", () => {
  it("no status can move to new — checked per status, not only via the matrix", () => {
    for (const from of LEAD_STATUSES) {
      expect(isValidStatusTransition(from, "new"), `${from} -> new`).toBe(false);
    }
  });

  it("the UI never offers a way into new, from any status", () => {
    for (const from of LEAD_STATUSES) {
      const targets = statusTransitionOptions(from).map((o) => o.status);
      expect(targets, `options from ${from}`).not.toContain("new");
    }
  });

  it('"contacted" offers forward and the side exits, but no backward step', () => {
    const steps = statusTransitionOptions("contacted").filter((o) => o.kind === "step");
    expect(steps.map((o) => o.status)).toEqual(["qualified"]);
  });
});

describe("statusTransitionOptions", () => {
  it("offers nothing for the two final statuses", () => {
    expect(statusTransitionOptions("unqualified")).toEqual([]);
    expect(statusTransitionOptions("spam")).toEqual([]);
  });

  it("offers only both main-path neighbors plus the three side exits for a middle status", () => {
    const options = statusTransitionOptions("qualified");
    expect(options.map((o) => o.status).sort()).toEqual(
      ["contacted", "proposal", "lost", "unqualified", "spam"].sort()
    );
  });

  it("offers only forward, no backward, for the first main-path status", () => {
    const options = statusTransitionOptions("new");
    const steps = options.filter((o) => o.kind === "step");
    expect(steps.map((o) => o.status)).toEqual(["contacted"]);
  });

  it("offers only backward, no forward, for the last main-path status", () => {
    const options = statusTransitionOptions("won");
    const steps = options.filter((o) => o.kind === "step");
    expect(steps.map((o) => o.status)).toEqual(["proposal"]);
  });

  it('restricts "lost" to a single reopen-to-"contacted" option, not all 5 valid main-path targets', () => {
    // isValidStatusTransition allows lost -> any main-path status except
    // new; the UI deliberately narrows this to avoid a wall of buttons for
    // a rare case. This test locks that narrowing as intentional — and the
    // target: "new" is closed (ADR-015), "contacted" is the earliest stage
    // a reopened lead can honestly be in.
    const options = statusTransitionOptions("lost");
    expect(options).toHaveLength(1);
    expect(options[0].status).toBe("contacted");
    expect(options[0].label).toBe("→ Buka kembali ke Dihubungi");
  });

  it("every option returned is actually valid per isValidStatusTransition", () => {
    for (const from of LEAD_STATUSES) {
      for (const option of statusTransitionOptions(from)) {
        expect(isValidStatusTransition(from, option.status)).toBe(true);
      }
    }
  });
});

// Issue #142. The timeline's "lead_converted" entry is the only "already
// converted" signal the client has (the Lead carries no such flag), so the
// predicate that drives "stop offering the status buttons" is worth pinning.
describe("hasBeenConverted", () => {
  it("is true once the timeline carries a lead_converted entry, wherever it sits", () => {
    expect(hasBeenConverted([{ type: "lead_created" }, { type: "lead_converted" }])).toBe(true);
    expect(hasBeenConverted([{ type: "lead_converted" }, { type: "status_changed" }])).toBe(true);
  });

  it("is false for a timeline without one — including a won lead that was never converted", () => {
    expect(hasBeenConverted([])).toBe(false);
    expect(
      hasBeenConverted([{ type: "lead_created" }, { type: "status_changed" }, { type: "note" }])
    ).toBe(false);
  });
});

File: crm_employee/test/features/leads/domain/lead_status_test.dart
import 'package:crm_employee/features/leads/domain/entities/activity.dart';
import 'package:crm_employee/features/leads/domain/lead_status.dart';
import 'package:crm_employee/shared/labels.dart';
import 'package:flutter_test/flutter_test.dart';

// This is the literal transition matrix from
// crm_be/internal/lead/usecase.go's validateStatusTransition, ported
// verbatim from crm_dashboard/src/lib/lead-status.test.ts (itself worked
// out by hand from the Go source, not derived from
// isValidStatusTransition) — so a bug that breaks BOTH ports the same
// way still can't hide.
//
// Nothing leads to 'new' (issue #139, ADR-015): it means "belum disentuh",
// a statement about history rather than a stage of work, so no row below
// lists it — not even lost's, which is why reopening goes to 'contacted'.
const Map<String, List<String>> _expectedValid = {
  'new': ['contacted', 'lost', 'unqualified', 'spam'],
  'contacted': ['qualified', 'lost', 'unqualified', 'spam'],
  'qualified': ['contacted', 'proposal', 'lost', 'unqualified', 'spam'],
  'proposal': ['qualified', 'won', 'lost', 'unqualified', 'spam'],
  'won': ['proposal', 'lost', 'unqualified', 'spam'],
  // "leaving lost" is a documented backend simplification: ANY main-path
  // status is valid, not just the one the lead was in before — there's
  // no cheap way to know "before" without activity history (crm_be #20
  // notes). unqualified/spam are also reachable directly from lost.
  'lost': [
    'contacted',
    'qualified',
    'proposal',
    'won',
    'unqualified',
    'spam',
  ],
  'unqualified': [],
  'spam': [],
};

void main() {
  group('isValidStatusTransition', () {
    test('matches the full matrix for every (from, to) pair — including same-status', () {
      for (final from in leadStatuses) {
        for (final to in leadStatuses) {
          final expected = to == from
              ? to == 'lost'
              : _expectedValid[from]!.contains(to);
          expect(
            isValidStatusTransition(from, to),
            expected,
            reason: '$from -> $to expected $expected',
          );
        }
      }
    });

    test('unqualified and spam are final — no outgoing transition at all, not even to themselves', () {
      for (final to in leadStatuses) {
        expect(isValidStatusTransition('unqualified', to), isFalse);
        expect(isValidStatusTransition('spam', to), isFalse);
      }
    });

    test('main path movement is exactly one step in either direction', () {
      expect(isValidStatusTransition('qualified', 'contacted'), isTrue); // back
      expect(isValidStatusTransition('qualified', 'proposal'), isTrue); // forward
      expect(isValidStatusTransition('proposal', 'qualified'), isTrue); // back
      expect(isValidStatusTransition('won', 'proposal'), isTrue); // back
      expect(isValidStatusTransition('proposal', 'new'), isFalse); // two steps back
      expect(isValidStatusTransition('new', 'won'), isFalse); // skips ahead
    });
  });

  group('nothing leads back to new (issue #139)', () {
    test('no status can move to new — checked per status, not only via the matrix', () {
      for (final from in leadStatuses) {
        expect(
          isValidStatusTransition(from, 'new'),
          isFalse,
          reason: '$from -> new',
        );
      }
    });

    test('the UI never offers a way into new, from any status', () {
      for (final from in leadStatuses) {
        final targets = statusTransitionOptions(from).map((o) => o.status);
        expect(targets, isNot(contains('new')), reason: 'options from $from');
      }
    });

    test('"contacted" offers forward and the side exits, but no backward step', () {
      final steps = statusTransitionOptions('contacted')
          .where((o) => o.kind == StatusTransitionKind.step)
          .map((o) => o.status);
      expect(steps, ['qualified']);
    });
  });

  group('statusTransitionOptions', () {
    test('offers nothing for the two final statuses', () {
      expect(statusTransitionOptions('unqualified'), isEmpty);
      expect(statusTransitionOptions('spam'), isEmpty);
    });

    test('offers only both main-path neighbors plus the three side exits for a middle status', () {
      final options = statusTransitionOptions('qualified');
      final statuses = options.map((o) => o.status).toList()..sort();
      expect(
        statuses,
        ['contacted', 'proposal', 'lost', 'unqualified', 'spam']..sort(),
      );
    });

    test('offers only forward, no backward, for the first main-path status', () {
      final options = statusTransitionOptions('new');
      final steps = options
          .where((o) => o.kind == StatusTransitionKind.step)
          .map((o) => o.status);
      expect(steps, ['contacted']);
    });

    test('offers only backward, no forward, for the last main-path status', () {
      final options = statusTransitionOptions('won');
      final steps = options
          .where((o) => o.kind == StatusTransitionKind.step)
          .map((o) => o.status);
      expect(steps, ['proposal']);
    });

    test('restricts "lost" to a single reopen-to-"contacted" option, not all valid main-path targets', () {
      // isValidStatusTransition allows lost -> any main-path status except
      // new; the UI deliberately narrows this to avoid a wall of buttons
      // for a rare case. This test locks that narrowing as intentional —
      // and the target: 'new' is closed (ADR-015), 'contacted' is the
      // earliest stage a reopened lead can honestly be in.
      final options = statusTransitionOptions('lost');
      expect(options, hasLength(1));
      expect(options[0].status, 'contacted');
      expect(options[0].label, '→ Buka kembali ke Dihubungi');
    });

    test('every option returned is actually valid per isValidStatusTransition — AC #72: UI never offers what the backend rejects', () {
      for (final from in leadStatuses) {
        for (final option in statusTransitionOptions(from)) {
          expect(isValidStatusTransition(from, option.status), isTrue);
        }
      }
    });
  });

  // Issue #142. The timeline's `lead_converted` entry is the only "already
  // converted" signal the client has (the Lead carries no such flag), so the
  // predicate that drives "stop offering the status picker" is worth pinning.
  group('hasBeenConverted', () {
    Activity entry(String type) => Activity(
          id: 'a-$type',
          leadId: 'lead-1',
          type: type,
          createdAt: DateTime.utc(2026, 9, 21),
        );

    test('is true once the timeline carries a lead_converted entry, wherever it sits', () {
      expect(hasBeenConverted([entry('lead_created'), entry('lead_converted')]), isTrue);
      expect(hasBeenConverted([entry('lead_converted'), entry('status_changed')]), isTrue);
    });

    test('is false for a timeline without one — including a won lead that was never converted', () {
      expect(hasBeenConverted(const []), isFalse);
      expect(
        hasBeenConverted([entry('lead_created'), entry('status_changed'), entry('note')]),
        isFalse,
      );
    });
  });
}
